sql/schema.go

```diff
@@ -0,0 +1,202 @@
+package sql
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	godiffpatch "github.com/sourcegraph/go-diff-patch"
+	"go.uber.org/zap"
+)
+
+const (
+	SchemaPath        = "schema/schema.sql"
+	UpdatedSchemaPath = "schema/schema.sql.updated"
+)
+
+// LoadSchema loads the schema embedded in the executable.
+func LoadSchema(fsys fs.FS, migrations []Migration) (*Schema, error) {
+	text, err := fs.ReadFile(fsys, SchemaPath)
+	if err != nil {
+		return nil, fmt.Errorf("error reading schema file %s: %w", SchemaPath, err)
+	}
+	return &Schema{Script: string(text), Migrations: migrations}, nil
+}
+
+// LoadDBSchemaScript retrieves the database schema as text.
+func LoadDBSchemaScript(db Executor, ignoreRx string) (string, error) {
+	var (
+		err   error
+		ignRx *regexp.Regexp
+		sb    strings.Builder
+	)
+	if ignoreRx != "" {
+		ignRx, err = regexp.Compile(ignoreRx)
+		if err != nil {
+			return "", fmt.Errorf("error compiling table ignore regexp %q: %w",
+				ignoreRx, err)
+		}
+	}
+	version, err := version(db)
+	if err != nil {
+		return "", err
+	}
+	fmt.Fprintf(&sb, "PRAGMA user_version = %d;\n", version)
+	if _, err = db.Exec(
+		// Type is either 'index' or 'table', we want tables
+		// to go first. Also, we ignore _litestream tables
+		`select tbl_name, sql || ';' from sqlite_master
+                 where sql is not null
+                 order by tbl_name, type desc, name`,
+		nil, func(st *Statement) bool {
+			if ignRx == nil || !ignRx.MatchString(st.ColumnText(0)) {
+				fmt.Fprintln(&sb, st.ColumnText(1))
+			}
+			return true
+		}); err != nil {
+		return "", fmt.Errorf("error retrieving DB schema: %w", err)
+	}
+	return sb.String(), nil
+}
+
+// Schema represents database schema.
+type Schema struct {
+	Script        string
+	Migrations    MigrationList
+	skipMigration map[int]struct{}
+}
+
+// Diff diffs the database schema against the actual schema.
+// If there's no differences, it returns an empty string.
+func (s *Schema) Diff(actualScript string) string {
+	if s.Script == actualScript {
+		return ""
+	}
+	diff := godiffpatch.GeneratePatch(SchemaPath, s.Script, actualScript)
+	if diff == "" {
+		return "<diff>"
+	}
+	return diff
```

Comment:
I don't understand this part. Why not return the diff as it is? I'd assume that an empty `diff` means that schemas are equal.
```suggestion
	return godiffpatch.GeneratePatch(SchemaPath, s.Script, actualScript)
```

Also, why bother generating a diff at all? Is it for debugging purposes?

Reply:
> I'd assume that an empty `diff` means that schemas are equal.

Well I'd assume that too, but wanted to be on the safe side, in case if `GeneratePatch` doesn't work as expected.
Maybe it would be more appropriate to `panic` here, need to think about it.

> Also, why bother generating a diff at all? Is it for debugging purposes?

Yes, in case there are differences such as some index I've created by hand, an extra table column or a leftover from running an intermediate unreleased version it makes things much easier to see exactly what's the difference between the expected schema and the actual one.

Reply:
> Maybe it would be more appropriate to panic here, need to think about it.

I don't think so? Does every call to `Diff` expect a difference between the actual script and what it is supposed to be? Also the log in `sql/database.go` is never called because this function doesn't ever return `""`.

Reply:
> Yes, in case there are differences such as some index I've created by hand, an extra table column or a leftover from running an intermediate unreleased version it makes things much easier to see exactly what's the difference between the expected schema and the actual one.

In that case I would log the diff as `Debug` log in `sql.Open` and return an error like `errors.New("open DB: schema drift")`, which should exit the node gracefully if that happens during startup and cause a panic when using `InMemory`.

Reply:
Also: we shouldn't introduce a dependency just for logging a diff of two schemas. We already have for instance `github.com/google/go-cmp` in our dependencies that does basically the same as `godiffpatch`

Reply:
> I don't think so? Does every call to `Diff` expect a difference between the actual script and what it is supposed to be? Also the log in `sql/database.go` is never called because this function doesn't ever return `""`.

There's a check above that diff that checks if the schema has changes
```
	if s.Script == actualScript {
		return ""
	}
```
So diff is expected to return a non-empty string here

Reply:
> In that case I would log the diff as `Debug` log in `sql.Open` and return an error like `errors.New("open DB: schema drift")`, which should exit the node gracefully if that happens during startup and cause a panic when using `InMemory`.

It is often very useful to allow schema drift while debugging go-spacemesh. For example, I might want to check if there's some performance improvement if I create some index, and that also constitutes an event of schema drift. We could of course have a config key for tolerating schema drift, but I'm not 100% sure I've anticipated all the situations which might cause adding extra stuff to the database, so I'm not sure if there will be no complaints about go-sm not working

Reply:
Switched to go-cmp

Reply:
> It is often very useful to allow schema drift while debugging go-spacemesh. For example, I might want to check if there's some performance improvement if I create some index, and that also constitutes an event of schema drift. We could of course have a config key for tolerating schema drift, but I'm not 100% sure I've anticipated all the situations which might cause adding extra stuff to the database, so I'm not sure if there will be no complaints about go-sm not working

Alternatively the check that returns the error if the diff isn't empty can also be commented out during debugging 😉 I don't see a reason to allow such a thing and build a node that actively has migration issues to then deploy it?

Reply:
If you think that it makes sense however to allow drifts in some cases I would go for the configuration option - if drift is allowed log the warning as now, if it is not return an error which will shut down the node gracefully.

Reply:
Schema drift is now fatal by default